lib/system/gc_interface.nim

```diff
@@ -13,6 +13,23 @@ when hasAlloc:
       gcOptimizeTime,    ## optimize for speed
       gcOptimizeSpace    ## optimize for memory footprint
 
+type
+  GCStatsData* = object
```

Comment:
I'm not sure if we should use an object like this. Not all supported GC even have some of the concepts in this object, let alone data for it.

I think what should be done instead is to define a set of data querying operations, like `GC_getMemoryUsage`, `GC_getCycleStats`, etc. and expose only the interfaces supported by the GC in use.

We should then be able to check if a given statistic is available via something like:

```nim
when declared(GC_getCycleStats):
  let cycleStats = GC_getCycleStats()
  <do stuff with this data>
```

An alternative is to use a lot of `when` in the `GCStats` object, for example:

```nim
type GCStats = object
  # Common data goes here
  totalMem*: int ## Total amount of memory allocated from the OS
  occupiedMem*: int ## The amount of memory in use
  when defined(gcRefc) or defined(gcMarkAndSweep) or defined(gcOrc):
    cycleCollections*: int    ## number of performed full collections (orc, refc, markAndSweep only)

  when defined(gcRefc) or defined(gcMarkAndSweep):
    cycleTableSize*: int      ## max entries in cycle table (refc, markAndSweep only)
    maxCycleTableSize*: int64 ## Only available on refc, markAndSweep GCs.
    maxPause*: int64          ## Only available on refc, markAndSweep GCs.
    maxPauseTime*: int64      ## Only available on refc, markAndSweep GCs.
    maxStackCells*: int       ## Only available on refc, markAndSweep GCs.
```

Why `when`s? It's because you cannot change the GC on-the-fly, so it makes little sense to have all the data for cycle collections when you use arc, for example.

Reply:
You can't choose gc on the fly, but you certainly can try to analyze the data from two program runs with different compiler, and for that (if we dumped data in a structured manner) we need all fields to be accessible via variant object that switches on used gc type.